File: 99_visualisations/themes.R
# ggplot2 Themes for Clinical Flow Cytometry Gender Analysis
# ============================================================
#
# This script provides custom ggplot2 themes that enforce the project style guide.
# Apply these themes at the start of plotting scripts to ensure consistency.
#
# Author: Lucas Black
# Date: 2025-10-28
# Version: 1.0

library(ggplot2)

# =============================================================================
# FIGURE DIMENSIONS
# =============================================================================

FIGURE_WIDTH <- 13.33  # inches
FIGURE_HEIGHT <- 7.5   # inches
FIGURE_DPI <- 150

# Multi-panel figure dimensions
FIGURE_2x1_WIDTH <- 13.33
FIGURE_2x1_HEIGHT <- 7.5
FIGURE_1x2_WIDTH <- 13.33
FIGURE_1x2_HEIGHT <- 7.5
FIGURE_2x2_WIDTH <- 13.33
FIGURE_2x2_HEIGHT <- 7.5


# =============================================================================
# FONT SIZES
# =============================================================================

FONT_SIZE_AXIS_TITLE <- 14
FONT_SIZE_AXIS_LABEL <- 12
FONT_SIZE_LEGEND_TITLE <- 12
FONT_SIZE_LEGEND_TEXT <- 10
FONT_SIZE_PLOT_TITLE <- 16
FONT_SIZE_ANNOTATION <- 10
FONT_SIZE_PANEL_LABEL <- 18


# =============================================================================
# PROJECT THEME
# =============================================================================

#' Apply project-wide ggplot2 theme
#'
#' This creates a custom theme that matches the style guide specifications.
#' Call this at the start of any plotting script using theme_set().
#'
#' @param base_size Base font size (default: 12)
#' @param base_family Base font family (default: "Arial")
#' @return A ggplot2 theme object
#'
#' @examples
#' library(ggplot2)
#' theme_set(theme_project())
#' # Now all plots will use project theme
theme_project <- function(base_size = 12, base_family = "Arial") {
  theme_minimal(base_size = base_size, base_family = base_family) +
    theme(
      # Axis titles (bold, size 14)
      axis.title = element_text(
        size = FONT_SIZE_AXIS_TITLE,
        face = "bold",
        color = "black"
      ),

      # Axis text (size 12)
      axis.text = element_text(
        size = FONT_SIZE_AXIS_LABEL,
        color = "black"
      ),

      # Plot title (bold, size 16, centered)
      plot.title = element_text(
        size = FONT_SIZE_PLOT_TITLE,
        face = "bold",
        hjust = 0.5,
        color = "black"
      ),

      # Legend title (bold, size 12)
      legend.title = element_text(
        size = FONT_SIZE_LEGEND_TITLE,
        face = "bold",
        color = "black"
      ),

      # Legend text (size 10)
      legend.text = element_text(
        size = FONT_SIZE_LEGEND_TEXT,
        color = "black"
      ),

      # Legend position (bottom by default, can override)
      legend.position = "bottom",

      # Legend background
      legend.background = element_rect(
        fill = "white",
        color = "#CCCCCC",
        size = 0.5
      ),

      # Panel grid (no minor grid, major grid dashed)
      panel.grid.minor = element_blank(),
      panel.grid.major = element_line(
        color = "gray90",
        linetype = "dashed",
        size = 0.5
      ),

      # Panel background
      panel.background = element_rect(fill = "white", color = NA),

      # Plot background
      plot.background = element_rect(fill = "white", color = NA),

      # Facet strip text (for multi-panel plots)
      strip.text = element_text(
        size = FONT_SIZE_AXIS_LABEL,
        face = "bold",
        color = "black"
      ),

      strip.background = element_rect(
        fill = "#F0F0F0",
        color = "#CCCCCC",
        size = 0.5
      )
    )
}


#' Set project theme globally
#'
#' Convenience function to set project theme for all subsequent plots.
#'
#' @examples
#' set_project_theme()
#' # All plots now use project theme
set_project_theme <- function() {
  theme_set(theme_project())
  message("Project theme applied successfully.")
}


#' Reset to default ggplot2 theme
#'
#' @examples
#' reset_theme()
reset_theme <- function() {
  theme_set(theme_gray())
  message("ggplot2 reset to default theme.")
}


# =============================================================================
# CONVENIENCE FUNCTIONS
# =============================================================================

#' Save figure in multiple formats following project standards
#'
#' @param plot ggplot object to save
#' @param filename Base filename (without extension)
#' @param width Figure width in inches (default: 13.33)
#' @param height Figure height in inches (default: 7.5)
#' @param dpi Resolution (default: 150)
#' @param formats Vector of formats to save (default: c("png", "pdf"))
#'
#' @examples
#' p <- ggplot(mtcars, aes(mpg, wt)) + geom_point()
#' save_figure(p, "figures/fig1_example_v1")
save_figure <- function(plot,
                        filename,
                        width = FIGURE_WIDTH,
                        height = FIGURE_HEIGHT,
                        dpi = FIGURE_DPI,
                        formats = c("png", "pdf")) {

  for (fmt in formats) {
    output_path <- paste0(filename, ".", fmt)
    ggsave(
      output_path,
      plot = plot,
      width = width,
      height = height,
      dpi = dpi,
      units = "in",
      bg = "white"
    )
    message("Saved: ", output_path)
  }
}


#' Add panel label to plot
#'
#' @param label Panel label (e.g., "A", "B", "C")
#' @param x X position in npc coordinates (default: 0.02)
#' @param y Y position in npc coordinates (default: 0.98)
#' @param size Font size (default: 18)
#' @param fontface Font face (default: "bold")
#' @return A geom_text layer
#'
#' @examples
#' library(ggplot2)
#' ggplot(mtcars, aes(mpg, wt)) +
#'   geom_point() +
#'   add_panel_label("A")
add_panel_label <- function(label,
                             x = 0.02,
                             y = 0.98,
                             size = FONT_SIZE_PANEL_LABEL,
                             fontface = "bold") {
  annotate(
    "text",
    x = x,
    y = y,
    label = label,
    size = size / .pt,  # Convert to mm
    fontface = fontface,
    hjust = 0,
    vjust = 1
  )
}


#' Add sample size text to plot
#'
#' @param text Sample size text (e.g., "F: n = 120,000 | M: n = 150,000")
#' @param x X position in npc coordinates (default: 0.5, centered)
#' @param y Y position in npc coordinates (default: 0.95)
#' @param size Font size (default: 10)
#' @param color Text color (default: "#4B5563")
#' @return An annotation layer
#'
#' @examples
#' library(ggplot2)
#' ggplot(mtcars, aes(mpg, wt)) +
#'   geom_point() +
#'   add_sample_size_text("n = 32")
add_sample_size_text <- function(text,
                                  x = 0.5,
                                  y = 0.95,
                                  size = FONT_SIZE_ANNOTATION,
                                  color = "#4B5563") {
  annotate(
    "text",
    x = x,
    y = y,
    label = text,
    size = size / .pt,
    fontface = "italic",
    color = color,
    hjust = 0.5,
    vjust = 1
  )
}


#' Add significance bracket with star notation
#'
#' @param x1 X position of left end (data coordinates)
#' @param x2 X position of right end (data coordinates)
#' @param y Y position of bracket (data coordinates)
#' @param text Significance text (e.g., "***", "**", "*", "ns")
#' @param size Font size (default: 10)
#' @return A list of annotation layers
#'
#' @examples
#' library(ggplot2)
#' ggplot(mtcars, aes(factor(cyl), mpg)) +
#'   geom_boxplot() +
#'   add_significance_bracket(1, 2, 35, "***")
add_significance_bracket <- function(x1,
                                      x2,
                                      y,
                                      text,
                                      size = FONT_SIZE_ANNOTATION) {
  list(
    # Bracket line
    annotate(
      "segment",
      x = x1,
      xend = x2,
      y = y,
      yend = y,
      color = "black",
      size = 0.5
    ),
    # Text
    annotate(
      "text",
      x = (x1 + x2) / 2,
      y = y,
      label = text,
      size = size / .pt,
      vjust = -0.5,
      hjust = 0.5
    )
  )
}


#' Enable/disable grid on specific axis
#'
#' @param axis Which axis: "y", "x", or "both" (default: "y")
#' @return A theme modification
#'
#' @examples
#' library(ggplot2)
#' ggplot(mtcars, aes(mpg, wt)) +
#'   geom_point() +
#'   enable_grid("y")
enable_grid <- function(axis = "y") {
  if (axis == "y") {
    theme(
      panel.grid.major.y = element_line(
        color = "gray90",
        linetype = "dashed",
        size = 0.5
      ),
      panel.grid.major.x = element_blank()
    )
  } else if (axis == "x") {
    theme(
      panel.grid.major.x = element_line(
        color = "gray90",
        linestyle = "dashed",
        size = 0.5
      ),
      panel.grid.major.y = element_blank()
    )
  } else if (axis == "both") {
    theme(
      panel.grid.major = element_line(
        color = "gray90",
        linetype = "dashed",
        size = 0.5
      )
    )
  }
}


# =============================================================================
# EXAMPLE USAGE
# =============================================================================

if (interactive()) {
  library(ggplot2)

  # Apply project theme
  set_project_theme()

  # Source color palettes
  source("color_palettes.R")

  # Create example plot
  example_data <- data.frame(
    gender = c("F", "M"),
    centrality = c(12.4, 15.1),
    n = c(120000, 150000)
  )

  p <- ggplot(example_data, aes(x = gender, y = centrality, fill = gender)) +
    geom_bar(stat = "identity") +
    scale_fill_manual(
      values = gender_colors,
      labels = gender_labels
    ) +
    scale_x_discrete(labels = c("F" = "Female", "M" = "Male")) +
    labs(
      x = "Inferred Gender",
      y = "Mean Degree Centrality",
      title = "Example Plot with Project Theme",
      fill = "Inferred Gender"
    ) +
    add_sample_size_text("F: n = 120,000 | M: n = 150,000") +
    add_significance_bracket(1, 2, max(example_data$centrality) * 1.1, "***") +
    enable_grid("y")

  print(p)

  # Save figure
  save_figure(p, "example_plot_with_theme_R")

  cat("\nExample plot created successfully!\n")
  cat("Check example_plot_with_theme_R.png and example_plot_with_theme_R.pdf\n")
}

File: 99_visualisations/themes.py
"""
Matplotlib Themes for Clinical Flow Cytometry Gender Analysis
==============================================================

This module provides custom matplotlib themes that enforce the project style guide.
Apply these themes at the start of plotting scripts to ensure consistency.

Author: Lucas Black
Date: 2025-10-28
Version: 1.0
"""

import matplotlib.pyplot as plt
import matplotlib as mpl


# =============================================================================
# FIGURE DIMENSIONS
# =============================================================================

FIGURE_WIDTH = 13.33  # inches
FIGURE_HEIGHT = 7.5   # inches
FIGURE_DPI = 150

# Multi-panel figure dimensions
FIGURE_2x1_WIDTH = 13.33
FIGURE_2x1_HEIGHT = 7.5
FIGURE_1x2_WIDTH = 13.33
FIGURE_1x2_HEIGHT = 7.5
FIGURE_2x2_WIDTH = 13.33
FIGURE_2x2_HEIGHT = 7.5


# =============================================================================
# FONT SIZES
# =============================================================================

FONT_SIZE_AXIS_TITLE = 14
FONT_SIZE_AXIS_LABEL = 12
FONT_SIZE_LEGEND_TITLE = 12
FONT_SIZE_LEGEND_TEXT = 10
FONT_SIZE_PLOT_TITLE = 16
FONT_SIZE_ANNOTATION = 10
FONT_SIZE_PANEL_LABEL = 18


# =============================================================================
# PROJECT THEME
# =============================================================================

def set_project_theme():
    """
    Apply project-wide matplotlib theme.

    This sets all rcParams to match the style guide specifications.
    Call this at the start of any plotting script.

    Example
    -------
    >>> from themes import set_project_theme
    >>> set_project_theme()
    >>> fig, ax = plt.subplots()
    >>> # ... plotting code ...
    """

    # Font settings
    plt.rcParams['font.family'] = 'Arial'
    plt.rcParams['font.sans-serif'] = ['Arial', 'Helvetica', 'DejaVu Sans']
    plt.rcParams['font.size'] = 12

    # Font sizes for specific elements
    plt.rcParams['axes.titlesize'] = FONT_SIZE_AXIS_TITLE
    plt.rcParams['axes.labelsize'] = FONT_SIZE_AXIS_TITLE
    plt.rcParams['axes.titleweight'] = 'bold'
    plt.rcParams['axes.labelweight'] = 'bold'

    plt.rcParams['xtick.labelsize'] = FONT_SIZE_AXIS_LABEL
    plt.rcParams['ytick.labelsize'] = FONT_SIZE_AXIS_LABEL

    plt.rcParams['legend.fontsize'] = FONT_SIZE_LEGEND_TEXT
    plt.rcParams['legend.title_fontsize'] = FONT_SIZE_LEGEND_TITLE

    plt.rcParams['figure.titlesize'] = FONT_SIZE_PLOT_TITLE
    plt.rcParams['figure.titleweight'] = 'bold'

    # Figure settings
    plt.rcParams['figure.figsize'] = (FIGURE_WIDTH, FIGURE_HEIGHT)
    plt.rcParams['figure.dpi'] = FIGURE_DPI
    plt.rcParams['savefig.dpi'] = FIGURE_DPI
    plt.rcParams['savefig.bbox'] = 'tight'

    # Axes settings
    plt.rcParams['axes.spines.top'] = False
    plt.rcParams['axes.spines.right'] = False
    plt.rcParams['axes.linewidth'] = 1.0
    plt.rcParams['axes.edgecolor'] = 'black'

    # Grid settings
    plt.rcParams['axes.grid'] = False  # Enable per-plot as needed
    plt.rcParams['grid.alpha'] = 0.3
    plt.rcParams['grid.linestyle'] = '--'
    plt.rcParams['grid.linewidth'] = 0.5

    # Legend settings
    plt.rcParams['legend.frameon'] = True
    plt.rcParams['legend.framealpha'] = 0.8
    plt.rcParams['legend.edgecolor'] = '#CCCCCC'
    plt.rcParams['legend.fancybox'] = True

    # Line and marker settings
    plt.rcParams['lines.linewidth'] = 1.5
    plt.rcParams['lines.markersize'] = 6

    # Save settings
    plt.rcParams['savefig.format'] = 'png'
    plt.rcParams['savefig.transparent'] = False
    plt.rcParams['savefig.facecolor'] = 'white'

    print("Project theme applied successfully.")


def reset_theme():
    """
    Reset matplotlib to default settings.
    """
    mpl.rcdefaults()
    print("Matplotlib reset to defaults.")


# =============================================================================
# CONVENIENCE FUNCTIONS
# =============================================================================

def create_figure(nrows=1, ncols=1, **kwargs):
    """
    Create a figure with project defaults.

    Parameters
    ----------
    nrows : int, optional
        Number of subplot rows (default: 1)
    ncols : int, optional
        Number of subplot columns (default: 1)
    **kwargs
        Additional arguments passed to plt.subplots()

    Returns
    -------
    fig, ax or axes
        Matplotlib figure and axis/axes objects

    Example
    -------
    >>> fig, ax = create_figure()
    >>> fig, axes = create_figure(2, 2)
    """
    # Set default figsize if not provided
    if 'figsize' not in kwargs:
        kwargs['figsize'] = (FIGURE_WIDTH, FIGURE_HEIGHT)

    if 'dpi' not in kwargs:
        kwargs['dpi'] = FIGURE_DPI

    fig, ax = plt.subplots(nrows, ncols, **kwargs)
    return fig, ax


def save_figure(fig, filename, dpi=None, formats=['png', 'pdf']):
    """
    Save figure in multiple formats following project standards.

    Parameters
    ----------
    fig : matplotlib.figure.Figure
        Figure to save
    filename : str
        Base filename (without extension)
    dpi : int, optional
        Resolution (default: uses project DPI)
    formats : list of str, optional
        List of formats to save (default: ['png', 'pdf'])

    Example
    -------
    >>> save_figure(fig, 'figures/fig1_corpus_overview_v1')
    # Saves fig1_corpus_overview_v1.png and fig1_corpus_overview_v1.pdf
    """
    if dpi is None:
        dpi = FIGURE_DPI

    for fmt in formats:
        output_path = f"{filename}.{fmt}"
        fig.savefig(output_path, dpi=dpi, bbox_inches='tight', facecolor='white')
        print(f"Saved: {output_path}")


def add_panel_label(ax, label, x=0.02, y=0.98, **kwargs):
    """
    Add panel label (A, B, C, D) to subplot.

    Parameters
    ----------
    ax : matplotlib.axes.Axes
        Axis to add label to
    label : str
        Panel label (e.g., 'A', 'B', 'C')
    x : float, optional
        X position in axis coordinates (default: 0.02)
    y : float, optional
        Y position in axis coordinates (default: 0.98)
    **kwargs
        Additional text properties

    Example
    -------
    >>> fig, axes = create_figure(1, 2)
    >>> add_panel_label(axes[0], 'A')
    >>> add_panel_label(axes[1], 'B')
    """
    default_kwargs = {
        'fontsize': FONT_SIZE_PANEL_LABEL,
        'fontweight': 'bold',
        'va': 'top',
        'ha': 'left',
        'transform': ax.transAxes
    }
    default_kwargs.update(kwargs)

    ax.text(x, y, label, **default_kwargs)


def add_sample_size_text(ax, text, x=0.5, y=0.95, **kwargs):
    """
    Add sample size text to plot.

    Parameters
    ----------
    ax : matplotlib.axes.Axes
        Axis to add text to
    text : str
        Sample size text (e.g., 'F: n = 120,000 | M: n = 150,000')
    x : float, optional
        X position in axis coordinates (default: 0.5, centered)
    y : float, optional
        Y position in axis coordinates (default: 0.95)
    **kwargs
        Additional text properties

    Example
    -------
    >>> add_sample_size_text(ax, 'F: n = 120,000 | M: n = 150,000')
    """
    default_kwargs = {
        'fontsize': FONT_SIZE_ANNOTATION,
        'style': 'italic',
        'color': '#4B5563',
        'ha': 'center',
        'va': 'top',
        'transform': ax.transAxes
    }
    default_kwargs.update(kwargs)

    ax.text(x, y, text, **default_kwargs)


def add_significance_bracket(ax, x1, x2, y, text, **kwargs):
    """
    Add significance bracket with star notation.

    Parameters
    ----------
    ax : matplotlib.axes.Axes
        Axis to add bracket to
    x1 : float
        X position of left end of bracket (data coordinates)
    x2 : float
        X position of right end of bracket (data coordinates)
    y : float
        Y position of bracket (data coordinates)
    text : str
        Significance text (e.g., '***', '**', '*', 'ns')
    **kwargs
        Additional line/text properties

    Example
    -------
    >>> add_significance_bracket(ax, 0, 1, ymax*1.05, '***')
    """
    # Draw bracket
    ax.plot([x1, x2], [y, y], 'k-', linewidth=1)

    # Add text
    text_kwargs = {
        'fontsize': FONT_SIZE_ANNOTATION,
        'ha': 'center',
        'va': 'bottom'
    }
    text_kwargs.update(kwargs)

    ax.text((x1 + x2) / 2, y, text, **text_kwargs)


def enable_grid(ax, axis='y'):
    """
    Enable grid on axis following project style.

    Parameters
    ----------
    ax : matplotlib.axes.Axes
        Axis to enable grid on
    axis : {'both', 'x', 'y'}, optional
        Which axis to apply grid to (default: 'y')

    Example
    -------
    >>> enable_grid(ax, axis='y')
    """
    ax.grid(axis=axis, alpha=0.3, linestyle='--', linewidth=0.5)


# =============================================================================
# EXAMPLE USAGE
# =============================================================================

if __name__ == "__main__":
    # Demonstrate theme application
    set_project_theme()

    # Create a simple example plot
    fig, ax = create_figure()

    # Example data
    categories = ['Female', 'Male']
    values = [12.4, 15.1]

    ax.bar(categories, values, color=['#6a00ff', '#FCA63F'])
    ax.set_xlabel('Inferred Gender')
    ax.set_ylabel('Mean Degree Centrality')
    ax.set_title('Example Plot with Project Theme')

    # Add sample size
    add_sample_size_text(ax, 'F: n = 120,000 | M: n = 150,000')

    # Add significance bracket
    add_significance_bracket(ax, 0, 1, max(values) * 1.1, '***')

    # Enable grid
    enable_grid(ax, axis='y')

    # Save figure
    save_figure(fig, 'example_plot_with_theme')

    plt.show()

    print("\nExample plot created successfully!")
    print("Check example_plot_with_theme.png and example_plot_with_theme.pdf")
